lib/models/package-info-cache/package-info.js

```diff
@@ -415,6 +421,190 @@ class PackageInfo {
+
+      let addonPackageList = pkgInfo.discoverAddonAddons();
+      addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+    }
+
+    let addonPackageList = this.discoverProjectAddons();
+    addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+
+    this.bundledProjectAddons = projectAddons;
+
+    return projectAddons;
+  }
+
+  /**
+   * Create an instance of the addon represented by this packageInfo or (if we
+   * are supporting singletons and this is a singleton-enabled addon) check if
+   * we should be creating a proxy instead.
+   * NOTE: we assume that the value of 'isSingleton' does not change between
+   * calls to the constructor! A given addon is either a singleton or not for an
+   * entire run.
+   *
+   * @method getAddonInstance
+   * @param {} parent the addon/project that is to be the direct parent of the
+   * addon instance created here
+   * @param {*} project the project that is to contain this addon instance
+   * @return {Object} the constructed instance of the addon
+   */
+  getAddonInstance(parent, project) {
+    let ctor = this.getAddonConstructor();
+    let addon;
+
+    if (!isSupportingSingletons || !ctor.isSingleton) {
+      // We have a 'regular' addon or we're not supporting singleton addons at all,
+      // so just return a regular instance. It is not in any singleton cache.
+      addon = this.constructAddonInstance(parent, project);
+
+      if (addon.initializeAddons) {
+        addon.initializeAddons();
+      } else {
+        addon.addons = [];
+      }
+
+      return addon;
+    }
+
+    // Find the bundleRoot that will contain this singleton instance.
+    // Find some parent, if there is one, that defines the 'isBundleRoot' property and that
+    // returns a truthy value. 'Project' is always the ultimate bundle.
+
+    // An optimization that lazy engines does that we can do as well. If the bundle
+    // root is not the project, but the project does or is going to have an instance
+    // of this singleton in its own cache, then use the project cache.
+    let bundleRoot = this.findBundleRoot(parent);
+    let bundledProjectAddons = project.getPackageInfo().getBundledProjectAddons();
+
+    if (bundleRoot === project || (bundleRoot !== project && !bundledProjectAddons.has(this))) {
+      // the instance belongs to its bundle root, whether it's the project or not.
+      let cacheEntry = bundleRoot.singletonAddonCache.getEntry(this.realPath);
+
+      if (cacheEntry && cacheEntry.isBuilt) {
+        // Return a proxy for the already-set-up cache entry.
+        let proxy = getAddonProxy(cacheEntry, parent);
+        return proxy;
+      } else {
+        // The cache entry may or may not exist, but the instance isn't set up yet.
+        // Create a new instance, install it in the bundleRoot's cache (add/update
+        // the cache entry) and return the new instance.
+        addon = this.constructAddonInstance(parent, project);
+        if (ctor.isBundleRoot) {
```

Comment:
this should use the aforementioned util that @rwjblue recommended here: https://github.com/ember-cli/ember-cli/pull/9487/files#r598920189

Reply:
done.